### DISCUSSION TITLE
Foundation BOOLEAN

### DISCUSSION THREAD

**Discussion by Slion** [microsoft/CsWin32#584] [Q&A] [answered]
`HidD_GetProductString` return value is generated as that `Foundation.BOOLEAN` class. Is there an elegant way to check its value? I tried getting TRUE and FALSE to generate but I don't think that works.

I ended up declaring something like that, still somewhat ugly though:
```cs
namespace Windows.Win32
{
    class K
    {
        public static readonly Foundation.BOOLEAN TRUE = new Foundation.BOOLEAN(1);
        public static readonly Foundation.BOOLEAN FALSE = new Foundation.BOOLEAN(0);
    }
}
```

**Comment by AArnott** [comment, marked as answer]
Oh, that's interesting. We have both `BOOL` and `BOOLEAN` types (of different lengths) in the metadata. We already take steps in CsWin32 to make `BOOL` feel as much like `bool` as possible. We can do the same for `BOOLEAN`.

**Reply by AArnott** [reply]
See #585

**Reply by Slion** [reply]
Thanks for the quick turnaround.

**Reply by RussKie** [reply]
@aarnott it'd be great to have `BOOL.TRUE` and `BOOL.FALSE` and/or having some helper methods like `bool IsTrue()` and `bool IsFalse()` though, as it could help with the clarity. It's not obvious that `BOOL` is convertable to `bool` (C#), and I can totally see how consumers would write something like:
```cs
if (PInvoke.ReturnBOOL() == new BOOL(true)) { ... }
```
or - which is equally unpleasant (e.g. https://github.com/dotnet/winforms/pull/7444#discussion_r923987545):
```cs
if (PInvoke.ReturnBOOL() == true) { ... }
```

The same applies to `BOOLEAN` type.

**Reply by AArnott** [reply]
Adding more public APIs because people don't know they don't need them feels like replacing bad syntax with more bad syntax. I don't know why someone would ever write `== true`, as that is _never_ required for _any_ C# type. I can imagine your first example though of `== new BOOL(true)`, though I haven't seen that in the wild.

I think what I'd rather do is start adding analyzers to CsWin32 and start flagging when people write syntax like that to advise them that they needn't do so. What do you think, @russkie?

**Reply by RussKie** [reply]
> adding analyzers to CsWin32 and start flagging when people write syntax like that to advise them that they needn't do so.

Sounds like a great idea :)

**Reply by AArnott** [reply]
Filed https://github.com/microsoft/CsWin32/issues/607 to track.
